components/script/dom/paintworkletglobalscope.rs

```diff
@@ -54,44 +84,223 @@ impl PaintWorkletGlobalScope {
+    fn invoke_a_paint_callback(&self,
+                               name: Atom,
+                               size: Size2D<Au>,
+                               sender: Sender<Result<Image, PaintWorkletError>>)
+    {
+        let width = size.width.to_px().abs() as u32;
+        let height = size.height.to_px().abs() as u32;
+        debug!("Invoking a paint callback {}({},{}).", name, width, height);
+
+        let cx = self.worklet_global.get_cx();
+        let _ac = JSAutoCompartment::new(cx, self.worklet_global.reflector().get_jsobject().get());
+
+        // TODO: Steps 1-2.1.
+        // Step 3.
+        let definition = match self.paint_definitions.borrow().get(&name) {
+            Some(defn) => defn.clone(),
+            None => {
+                // Step 2.2.
+                warn!("Drawing un-registered paint definition {}.", name);
+                let image = self.placeholder_image(width, height, [0x00, 0x00, 0xFF, 0xFF]);
+                let _ = sender.send(Ok(image));
+                return;
+            }
+        };
+
+        // Steps 4-5
+        rooted!(in(cx) let mut paint_instance = UndefinedValue());
+        match self.paint_class_instances.borrow_mut().entry(name.clone()) {
+            Entry::Occupied(entry) => paint_instance.set(entry.get().get()),
+            Entry::Vacant(entry) => {
+                // Step 5.1
+                if !definition.constructor_valid_flag.get() {
+                    debug!("Drawing invalid paint definition {}.", name);
+                    let image = self.placeholder_image(width, height, [0x00, 0x00, 0xFF, 0xFF]);
+                    let _ = sender.send(Ok(image));
+                    return;
+                }
+                // Step 5.2-5.3
+                let args = HandleValueArray::new();
+                rooted!(in(cx) let mut result = null_mut());
+                unsafe { Construct1(cx, definition.class_constructor.handle(), &args, result.handle_mut()); }
+                paint_instance.set(ObjectValue(result.get()));
+                if unsafe { JS_IsExceptionPending(cx) } {
+                    debug!("Paint constructor threw an exception {}.", name);
+                    unsafe { JS_ClearPendingException(cx); }
+                    definition.constructor_valid_flag.set(false);
+                    let image = self.placeholder_image(width, height, [0x00, 0x00, 0xFF, 0xFF]);
+                    let _ = sender.send(Ok(image));
+                    return;
+                }
+                // Step 5.4
+                entry.insert(Heap::new(paint_instance.get()));
+            }
+        };
+
+        // TODO: Steps 6-7
+        // Step 8
+        let rendering_context = PaintRenderingContext2D::new(self);
+
+        // Step 9
+        let paint_size = PaintSize::new(self, size);
+
+        // TODO: Step 10
+        // Steps 11-12
+        debug!("Invoking paint function {}.", name);
+        let args = unsafe { HandleValueArray::from_rooted_slice(&[
```

Comment:
I am ok with the current code as long as we store the slice in a binding that outlives the call to from_rooted_slice, otherwise we are storing a pointer to a temporary stack value that could be overwritten before we actually reach Call.